### PR TITLE
The general algorithm

### PR DESCRIPTION
**Pull request by crisely09**
There is probably a better way to select the conditions...

### CODE REVIEW

File: flik/linesearch/linesearch.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        Variables' current values.
+    func: callable
+        Function :math:`f(x)`.
+    grad: callable
+        Gradient of the function :math:`\nabla f(x)`.
+    direction: {np.ndarray(N,), float, int}
+        Direction vector for next step.
+    alpha: float
+        The current step length estimate
+    conditions: str or list of str
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: {str, list of str}

File: flik/linesearch/linesearch.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        Conditions to be satisfied by the new step length.
+        Valid options are: 'soft-wolfe', 'strong-wolfe' and 'armijo'
+    const1: float
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: {float, 1e-4}

File: flik/linesearch/linesearch.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        Constant c_1 for Wolfe/Armijo conditions
+    const2: float
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: {float, 1e-2}

File: flik/linesearch/linesearch.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        raise TypeError("The conditions should be given as a str or a list of str.")
+    if not isinstance(const1, float):
+        raise TypeError("Constant c_1 should be given as a float.")
+    if not isinstance(const2, float):
+        raise TypeError("Constant c_2 should be given as a float.")
+    if not 0 < const1 < const2 < 1.0:
+        raise ValueError("Constants c_1 and c_2  must meet the condition:"
+                         " 0 < c_1 < c_2 < 1.0")
+    # Check duplicates if there are any
+    if len(conditions) > 1:
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: if len(conditions) == len(set(conditions))

File: flik/linesearch/linesearch.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        con = [c for n, c in enumerate(conditions) if c not in conditions[:n]]
+        if con != conditions:
+            raise ValueError("One or more conditions are duplicated.")
+    # Check valid conditions
+    for i, condition in enumerate(conditions):
+        if condition not in ['soft-wolfe', 'strong-wolfe', 'armijo']:
+            raise ValueError("Condition %d+1 is not implemented." % i)
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: raise ValueError('Condition, {}, is not supported'.format(condition))
remove enumerate

File: flik/linesearch/linesearch.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            if 'wolfe' in condition:
+                if 'soft' in condition:
+                    strong = False
+                else:
+                    strong = True
+                if wolfe(grad, var, direction, alpha_new, strong_wolfe=strong):
+                    return alpha_new
+            elif condition == 'armijo':
+                if armijo(func, grad, var, direction, alpha_new):
+                    return alpha_new
```
Comment (kimt33) [inline flik/linesearch/linesearch.py]: The algorithm doesn't quite do what it says it'll do. It doesn't make sure that all of the conditions are satisfied. It just returns the first alpha that satisfies one of the conditions.

**Review by kimt33 — CHANGES_REQUESTED** [review]

**Comment by kimt33** [comment]
I've made the requested changes. It doesn't seem to converge though. I think it has to do with the alpha update (because it gets stuck at 0.5).

File: flik/linesearch/linesearch.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            elif condition == 'armijo':
+                output.append(armijo(func, grad, var, direction, alpha_new))
+        return all(output)
+
+    # Interpolate until alpha_new satisfies the conditions
+    alpha_new = 1.0
+    while True:
+        alpha_new = cubic_interpolation(var, func, grad, direction, alpha, alpha_new)
+        for condition in conditions:
+            if satisfies_conditions(alpha_new):
```
Comment (crisely09) [inline flik/linesearch/linesearch.py]: This is definitely a nicer way to do it.

**Comment by kimt33** [comment]
I'm merging this, even though the line search module is not complete. This way, we can have all the code available to us when we refactor the line search module.

**Pull request merged**